[SPARK-16272][core] Allow config values to reference conf, env, system props.

vanzin opened the pull request:
This allows configuration to be more flexible, for example, when the cluster does
not have a homogeneous configuration (e.g. packages are installed on different
paths in different nodes). By allowing one to reference the environment from
the conf, it becomes possible to work around those in certain cases.

As part of the implementation, ConfigEntry now keeps track of all "known" configs
(i.e. those created through the use of ConfigBuilder), since that list is used
by the resolution code. This duplicates some code in SQLConf, which could potentially
be merged with this now. It will also make it simpler to implement some missing
features such as filtering which configs show up in the UI or in event logs - which
are not part of this change.

Another change is in the way ConfigEntry reads config data; it now takes a string
map and a function that reads env variables, so that it can be called both from
SparkConf and SQLConf. This makes it so both places follow the same read path,
instead of having to replicate certain logic in SQLConf. There are still a
couple of methods in SQLConf that peek into fields of ConfigEntry directly,
though.

Tested via unit tests, and by using the new variable expansion functionality
in a shell session with a custom spark.sql.hive.metastore.jars value.

Commit:
[SPARK-16272][core] Allow config values to reference conf, env, system props.

This allows configuration to be more flexible when the cluster does not
have a homogeneous configuration (e.g. packages are installed on different
paths in different nodes). By allowing one to reference the environment from
the conf, it becomes possible to work around those in certain cases.

The feature is hooked up to spark.sql.hive.metastore.jars to show how to use
it, and because it's an example of said scenario that I ran into. It uses a
new "pathConf" config type that is a shorthand for enabling variable expansion
on string configs.

As part of the implementation, ConfigEntry now keeps track of all "known" configs
(i.e. those created through the use of ConfigBuilder), since that list is used
by the resolution code. This duplicates some code in SQLConf, which could potentially
be merged with this now. It will also make it simpler to implement some missing
features such as filtering which configs show up in the UI or in event logs - which
are not part of this change.

Another change is in the way ConfigEntry reads config data; it now takes a string
map and a function that reads env variables, so that it can be called both from
SparkConf and SQLConf. This makes it so both places follow the same read path,
instead of having to replicate certain logic in SQLConf. There are still a
couple of methods in SQLConf that peek into fields of ConfigEntry directly,
though.

Tested via unit tests, and by using the new variable expansion functionality
in a shell session with a custom spark.sql.hive.metastore.jars value.

vanzin commented:
Note this contains the code for #14020 also since I needed that to test the changes.

SparkQA commented:
**[Test build #61643 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61643/consoleFull)** for PR 14022 at commit [`f4e772a`](https://github.com/apache/spark/commit/f4e772ad2b5c8231c48b2cd1f8f75c81e754c2b4).

SparkQA commented:
**[Test build #61643 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61643/consoleFull)** for PR 14022 at commit [`f4e772a`](https://github.com/apache/spark/commit/f4e772ad2b5c8231c48b2cd1f8f75c81e754c2b4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61643/
Test PASSed.

vanzin commented:
Randomly pinging a couple of people... @JoshRosen @srowen

srowen commented:
My only concern is the added complexity vs how useful it would be. Do we have weird corner cases here like circular references? or resolving references among many sources in the right order leading to odd ambiguity? maybe that's not possible and I haven't thought it through.

vanzin commented:
> Do we have weird corner cases here like circular references?

That's a good question; the current code doesn't handle that (if you reference a config that itself has references, those are not expanded), but it probably should for correctness. But implementing that creates the risk of circular references. Let me take a look, shouldn't be hard to solve.

re: multiple source, the prefix is mandatory, so there shouldn't be any ambiguity related to where the values come from. Also, only Spark config values that have been explicitly allowed do variable expansion.

re: usability, this is pretty much mandatory at least on YARN if the cluster doesn't have software installed on the same location on every node. That's more common than I'd like it to be, unfortunately.

Commit:
Allow chained references.

Expand values of known configs when doing variable expansion, but
detect and error out when circular references are found.

SparkQA commented:
**[Test build #61795 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61795/consoleFull)** for PR 14022 at commit [`240ab73`](https://github.com/apache/spark/commit/240ab73f155ada9a0e9e6374e2f3e9fb5bb02a34).

SparkQA commented:
**[Test build #61795 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61795/consoleFull)** for PR 14022 at commit [`240ab73`](https://github.com/apache/spark/commit/240ab73f155ada9a0e9e6374e2f3e9fb5bb02a34).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61795/
Test FAILed.

rxin commented:
Spark SQL allows `env:xxx` and `system:xxx`. We should follow the same here.

vanzin commented:
> Spark SQL allows env:xxx and system:xxx. We should follow the same here.

Sounds good. I looked briefly at the code and they could potentially be merged later, but to avoid issues like "how to support hiveconf:" I'll just modify this code to use "system:" and "env:" for the moment.

Commit:
Use SQL-style prefixes for the references.

"env:" and "system:", to match how SQL does variable expansion. We
can try to merge the two later, but since this code doesn't have
access to things like hadoop or hive configuration, let's leave them
separate for now.

SparkQA commented:
**[Test build #61924 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61924/consoleFull)** for PR 14022 at commit [`392bddc`](https://github.com/apache/spark/commit/392bddc57eaefb09c73902ea041f05705d9498aa).

SparkQA commented:
**[Test build #61924 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61924/consoleFull)** for PR 14022 at commit [`392bddc`](https://github.com/apache/spark/commit/392bddc57eaefb09c73902ea041f05705d9498aa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61924/
Test PASSed.

vanzin commented:
Any more comments? I'd like to get this is so I can work on cleaning up the code duplication between SQL and core.

Commit:
Merge branch 'master' into SPARK-16272

SparkQA commented:
**[Test build #62259 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62259/consoleFull)** for PR 14022 at commit [`b928a55`](https://github.com/apache/spark/commit/b928a55f855b66bc4ab0cb2f61cbc034e67dfcba).

SparkQA commented:
**[Test build #62259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62259/consoleFull)** for PR 14022 at commit [`b928a55`](https://github.com/apache/spark/commit/b928a55f855b66bc4ab0cb2f61cbc034e67dfcba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62259/
Test FAILed.

vanzin commented:
retest this please

SparkQA commented:
**[Test build #62275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62275/consoleFull)** for PR 14022 at commit [`b928a55`](https://github.com/apache/spark/commit/b928a55f855b66bc4ab0cb2f61cbc034e67dfcba).

SparkQA commented:
**[Test build #62275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62275/consoleFull)** for PR 14022 at commit [`b928a55`](https://github.com/apache/spark/commit/b928a55f855b66bc4ab0cb2f61cbc034e67dfcba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62275/
Test PASSed.

vanzin commented:
I'll leave this around a little more but if I don't see more feedback I plan to push it before the weekend.

ericl commented on core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala:
```diff
@@ -99,13 +118,83 @@ private class FallbackConfigEntry[T] (
+    val result = new StringBuilder()
+    var end = 0
+
+    while (end < value.length() && matcher.find(end)) {
```
I think you can use [Regex#replaceAllIn](http://daily-scala.blogspot.com/2010/02/regex-replaceallin.html) here instead of a loop.

vanzin replied:
That looks interesting, let me take a look.

ericl commented:
Instead of selectively enabling this for certain confs / config builders, how about a global flag for enabling config expansion? I think that would be less likely to be confusing. Also, perhaps a warning should be logged if there is a config value that looks expandable but the flag is off.

vanzin commented:
> how about a global flag for enabling config expansion

I think that would be more confusing. Why would someone disable expansion?

My only concern with enabling it for all options is performance - reading configs becomes more expensive. Maybe that's not a big deal and we can just do that.

ericl commented:
Wouldn't enabling it by default break backwards compatibility? I agree that would be better, but it seems likely that '${...}' may be used in existing configs.

vanzin commented:
> Wouldn't enabling it by default break backwards compatibility?

Yes, maybe. But having a flag to disable everything would also potentially break features that rely on it... although you could argue it's the user's fault at that point.

I just don't like adding new options that only exist because we can't make a decision about what their behavior should be. If variable substitution applies to all configs, then it does apply to all configs and you can't disable it. Since no configs that I know of have this functionality (the SQL variable substitution would be broken), from a user's perspective it's very unlikely doing it for all would break things.

ericl commented:
To reduce the risk, how about changing the semantics to

```
* - spark/sparkconf/hiveconf: looks for the value in the spark config
* - system: looks for the value in the system properties
* - env: looks for the value in the environment
* - (any other or no prefix): left as-is
```

vanzin commented:
That's mostly how it works; I would like to avoid an explicit `sparkconf:` prefix to avoid things like `sparkconf:spark.master`, but I can enforce that only variables starting with `spark.` are expanded easily.

ericl commented:
IMO `${sparkconf:spark.master}` is more clear to the unfamiliar reader, but it also seems ok to go with `${spark.master}`. Thought there might also be an issue if someone adds a spark conf that doesn't start with "spark.", which could be confusing.

Commit:
Apply substitution to all configs.

Also enforce a "spark." prefix when doing substitution.

Commit:
Use Regex.replaceAllIn.

SparkQA commented:
**[Test build #62339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62339/consoleFull)** for PR 14022 at commit [`cca8314`](https://github.com/apache/spark/commit/cca831428913a67836f1b363316e3d733acb1aea).

SparkQA commented:
**[Test build #62339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62339/consoleFull)** for PR 14022 at commit [`cca8314`](https://github.com/apache/spark/commit/cca831428913a67836f1b363316e3d733acb1aea).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62339/
Test PASSed.

vanzin commented:
> if someone adds a spark conf that doesn't start with "spark.", which I think is allowed by spark conf.

Well, it is and it isn't. You can't set non-`spark.` configs from spark-submit or system properties. But you can set them directly in the code. In general, Spark completely ignores anything that does not start with `spark.` so I think that's ok here.

ericl commented on core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala:
```diff
@@ -99,13 +133,66 @@ private class FallbackConfigEntry[T] (
+          }
+        case "system" => sys.props.get(name)
+        case "env" => Option(getenv(name))
+        case _ => throw new IllegalArgumentException(s"Invalid prefix: $prefix")
```
Should this throw?

vanzin replied:
Yeah, I have to take a look at this. Throwing might actually break some existing code in SparkHadoopUtil.

Commit:
Don't error out on unknown prefixes.

SparkQA commented:
**[Test build #62485 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62485/consoleFull)** for PR 14022 at commit [`6618bc4`](https://github.com/apache/spark/commit/6618bc4633c8e8663400f0afe86739d1021ca8f9).

Commit:
Drop "spark." requirement.

After thinking about it a bit more, it can lead to confusing behavior.
Since substitution only applies to configs declared internally by Spark,
this avoids random user configs being affected already, so no need for
another check to disable that.

SparkQA commented:
**[Test build #62488 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62488/consoleFull)** for PR 14022 at commit [`3afe221`](https://github.com/apache/spark/commit/3afe221114db035f4093ba2a6f703ea2c3382a3d).

ericl commented:
It looks good to me, except I really think you should keep the "spark." requirement. It seems plausible that some reason a string like '${x}' can show up in spark confs, perhaps things like spark.r.command.

Commit:
Revert "Drop "spark." requirement."

This reverts commit 3afe221114db035f4093ba2a6f703ea2c3382a3d.

vanzin commented:
>  I really think you should keep the "spark." requirement.

I reverted my last change since I forgot to change the test code anyway.

I'll spend some time to see if it's easy to handle SQLConf's WAREHOUSE_PATH with this code, to remove some manual variable expansion happening there, but if it's too much work I'll leave that for a later change.

SparkQA commented:
**[Test build #62489 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62489/consoleFull)** for PR 14022 at commit [`33de81f`](https://github.com/apache/spark/commit/33de81fdc41b77ab2d92ddab49bc4e3e74064fb9).

Commit:
Expand default values that are strings.

This allows some custom expansion code to be removed from SQLConf.

SparkQA commented:
**[Test build #62494 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62494/consoleFull)** for PR 14022 at commit [`ed5c18b`](https://github.com/apache/spark/commit/ed5c18baddbd7ceb4157f5a31bf150d2ef9e7d19).

SparkQA commented:
**[Test build #62485 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62485/consoleFull)** for PR 14022 at commit [`6618bc4`](https://github.com/apache/spark/commit/6618bc4633c8e8663400f0afe86739d1021ca8f9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62485/
Test PASSed.

SparkQA commented:
**[Test build #62488 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62488/consoleFull)** for PR 14022 at commit [`3afe221`](https://github.com/apache/spark/commit/3afe221114db035f4093ba2a6f703ea2c3382a3d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62488/
Test FAILed.

SparkQA commented:
**[Test build #62489 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62489/consoleFull)** for PR 14022 at commit [`33de81f`](https://github.com/apache/spark/commit/33de81fdc41b77ab2d92ddab49bc4e3e74064fb9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62489/
Test FAILed.

SparkQA commented:
**[Test build #62494 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62494/consoleFull)** for PR 14022 at commit [`ed5c18b`](https://github.com/apache/spark/commit/ed5c18baddbd7ceb4157f5a31bf150d2ef9e7d19).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62494/
Test PASSed.

vanzin commented:
I'll push this later after another test run. retest this please

SparkQA commented:
**[Test build #62637 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62637/consoleFull)** for PR 14022 at commit [`ed5c18b`](https://github.com/apache/spark/commit/ed5c18baddbd7ceb4157f5a31bf150d2ef9e7d19).

SparkQA commented:
**[Test build #62637 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62637/consoleFull)** for PR 14022 at commit [`ed5c18b`](https://github.com/apache/spark/commit/ed5c18baddbd7ceb4157f5a31bf150d2ef9e7d19).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62637/
Test PASSed.

vanzin commented:
Merging to master.
